state/cleanup.go

```diff
@@ -215,16 +215,41 @@ func (st *State) cleanupMachinesForDyingModel() (err error) {
 // cleanupServicesForDyingModel sets all services to Dying, if they are
 // not already Dying or Dead. It's expected to be used when a model is
 // destroyed.
-func (st *State) cleanupServicesForDyingModel() (err error) {
-	// This won't miss services, because a Dying model cannot have
-	// services added to it. But we do have to remove the services themselves
+func (st *State) cleanupApplicationsForDyingModel() (err error) {
+	if err := st.removeRemoteApplicationsForDyingModel(); err != nil {
+		return err
+	}
+	return st.removeApplicationsForDyingModel()
+}
+
+func (st *State) removeApplicationsForDyingModel() (err error) {
+	// This won't miss applications, because a Dying model cannot have
+	// applications added to it. But we do have to remove the applications themselves
 	// via individual transactions, because they could be in any state at all.
 	applications, closer := st.getCollection(applicationsC)
 	defer closer()
 	application := Application{st: st}
 	sel := bson.D{{"life", Alive}}
 	iter := applications.Find(sel).Iter()
-	defer closeIter(iter, &err, "reading service document")
+	defer closeIter(iter, &err, "reading application document")
+	for iter.Next(&application.doc) {
+		if err := application.Destroy(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (st *State) removeRemoteApplicationsForDyingModel() (err error) {
+	// This won't miss remote applications, because a Dying model cannot have
+	// applications added to it. But we do have to remove the applications themselves
+	// via individual transactions, because they could be in any state at all.
+	remoteApp, closer := st.getCollection(remoteApplicationsC)
+	defer closer()
+	application := RemoteApplication{st: st}
+	sel := bson.D{{"life", Alive}}
```

Comment:
extract all the code below out into a func? it's the same for the local and remote case.